dbr quant: insert activation obs explicitly, instead of relying on hooks

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #73493 dbr quant: enable reference module support for torch.qint32
* **#73492 dbr quant: insert activation obs explicitly, instead of relying on hooks**

Summary:

Before this PR, DBR quant reused the Eager mode quantization machinery
to insert activation observers. This was done for speed of developing the
prototype. A drawback of this is that the activation observers are not
present in DBR's data structures and live on the modules instead.

This PR refactors DBR quant to stop using Eager mode quantization
observer insertion for activations, and instead create and track the
activation observers in DBR's data structures. This has a couple of benefits:
1. activation observers are now created the same way in DBR for modules and functions
2. we can remove some technical debt due to fixing (1)
3. this will make it easier to support reference modules in a future PR

The reason (3) is true is because the current design of reference modules
assumes that the activation observer lives on the framework (like in FX
graph mode quantization). This PR starts to adhere to that assumption.

Test plan:

```
python test/test_quantization.py -k DBR
```

Differential Revision: [D34520758](https://our.internmc.facebook.com/intern/diff/D34520758)

vkuzo committed:
dbr quant: insert activation obs explicitly, instead of relying on hooks

Summary:

Before this PR, DBR quant reused the Eager mode quantization machinery
to insert activation observers. This was done for speed of developing the
prototype. A drawback of this is that the activation observers are not
present in DBR's data structures and live on the modules instead.

This PR refactors DBR quant to stop using Eager mode quantization
observer insertion for activations, and instead create and track the
activation observers in DBR's data structures. This has a couple of benefits:
1. activation observers are now created the same way in DBR for modules and functions
2. we can remove some technical debt due to fixing (1)
3. this will make it easier to support reference modules in a future PR

The reason (3) is true is because the current design of reference modules
assumes that the activation observer lives on the framework (like in FX
graph mode quantization). This PR starts to adhere to that assumption.

Test plan:

```
python test/test_quantization.py -k DBR
```

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34520758).

vkuzo committed:
Update on "dbr quant: insert activation obs explicitly, instead of relying on hooks"


Summary:

Before this PR, DBR quant reused the Eager mode quantization machinery
to insert activation observers. This was done for speed of developing the
prototype. A drawback of this is that the activation observers are not
present in DBR's data structures and live on the modules instead.

This PR refactors DBR quant to stop using Eager mode quantization
observer insertion for activations, and instead create and track the
activation observers in DBR's data structures. This has a couple of benefits:
1. activation observers are now created the same way in DBR for modules and functions
2. we can remove some technical debt due to fixing (1)
3. this will make it easier to support reference modules in a future PR

The reason (3) is true is because the current design of reference modules
assumes that the activation observer lives on the framework (like in FX
graph mode quantization). This PR starts to adhere to that assumption.

Test plan:

```
python test/test_quantization.py -k DBR
```

Differential Revision: [D34520758](https://our.internmc.facebook.com/intern/diff/D34520758)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34520758).

vkuzo commented:
@jerryzh168 , @andrewor14 , would you be able to help review this?

jerryzh168 commented on torch/ao/quantization/_quantize_dbr.py:
```diff
@@ -123,7 +137,7 @@ def convert(model: torch.nn.Module) -> torch.nn.Module:
     dynamic_mappings = get_default_dynamic_quant_module_mappings()
     # swap the modules
-    _swap_child_modules(model, static_mappings, dynamic_mappings)
+    _swap_child_modules(model, static_mappings, dynamic_mappings, None)
```
nit: can we add a default None for the last argument instead of passing None around?

jerryzh168 commented on torch/ao/quantization/_quantize_dbr.py:
```diff
@@ -104,11 +119,10 @@ def prepare(model, qconfig_dict, example_inputs, inplace=False, allow_list=None,
         elif isinstance(child, torch.nn.LSTM):
             # TODO: fix LSTM handling in eager mode static quant and remove this
-            child.qconfig = None
+            qconfig_dict['object_type'][torch.nn.LSTM] = None
```
qq: are we going to fix this with default custom_module support?

vkuzo replied:
The current logic in DBR for quantization of LSTM is just to make models with LSTMs not crash, it doesn't actually support LSTM quantization.  We would need a couple of PRs to add the support to the framework, it should be pretty straightforward.

vkuzo committed:
Update on "dbr quant: insert activation obs explicitly, instead of relying on hooks"


Summary:

Before this PR, DBR quant reused the Eager mode quantization machinery
to insert activation observers. This was done for speed of developing the
prototype. A drawback of this is that the activation observers are not
present in DBR's data structures and live on the modules instead.

This PR refactors DBR quant to stop using Eager mode quantization
observer insertion for activations, and instead create and track the
activation observers in DBR's data structures. This has a couple of benefits:
1. activation observers are now created the same way in DBR for modules and functions
2. we can remove some technical debt due to fixing (1)
3. this will make it easier to support reference modules in a future PR

The reason (3) is true is because the current design of reference modules
assumes that the activation observer lives on the framework (like in FX
graph mode quantization). This PR starts to adhere to that assumption.

Test plan:

```
python test/test_quantization.py -k DBR
```

Differential Revision: [D34520758](https://our.internmc.facebook.com/intern/diff/D34520758)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34520758).

vkuzo committed:
Update on "dbr quant: insert activation obs explicitly, instead of relying on hooks"


Summary:

Before this PR, DBR quant reused the Eager mode quantization machinery
to insert activation observers. This was done for speed of developing the
prototype. A drawback of this is that the activation observers are not
present in DBR's data structures and live on the modules instead.

This PR refactors DBR quant to stop using Eager mode quantization
observer insertion for activations, and instead create and track the
activation observers in DBR's data structures. This has a couple of benefits:
1. activation observers are now created the same way in DBR for modules and functions
2. we can remove some technical debt due to fixing (1)
3. this will make it easier to support reference modules in a future PR

The reason (3) is true is because the current design of reference modules
assumes that the activation observer lives on the framework (like in FX
graph mode quantization). This PR starts to adhere to that assumption.

Test plan:

```
python test/test_quantization.py -k DBR
```

Differential Revision: [D34520758](https://our.internmc.facebook.com/intern/diff/D34520758)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34520758).
